### PR TITLE
[Hue] Added support for sensor API

### LINKED ISSUE #2603: [Hue Binding]: add support for Sensor API
This issue is cloned from https://github.com/openhab/openhab2-addons/issues/1260 since Hue Binding is maintained in eclipse/smarthome repository.

Original issue description by @regnets

> I purchased a Philips Hue Dimmer (http://www2.meethue.com/de-de/productdetail/philips-hue-dimmer-switch) and this thing is not supported ootb from the hue binding as far as i can tell.
> 
> However the Philips API provides an interface for accessing those devices
> http://www.developers.meethue.com/documentation/sensors-api
> 
> Support for these devices would be great! However we will need a faster polling period for the sensors, as the API is still not providing any form of callback or external events.
> 

Additional description from me:

> With the appearance of Hue motion detector the support for Sensor API makes even more sense now. Even if implemented with polling. I guess it should work more or less fine from UX perspective for this new Hue motion detector which is in fact a multi-sensor providing: motion sensor, daylight sensor and (unadvertised) temperature sensor.
> 
> Since I got myself one of those, I would be up for the task :)
>

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added support for sensor API (Motion Sensor / Dimmer Switch)
- Removed dependency on `org.apache.commons`
- Do not export handlers anymore
- Fixed linter errors in documentation
- Minor improvements

This PR superseds #5996.

Closes #2603
Closes #5996

Testversion:
[org.eclipse.smarthome.binding.hue-0.10.0-SNAPSHOT.jar.zip](https://github.com/eclipse/smarthome/files/2567639/org.eclipse.smarthome.binding.hue-0.10.0-SNAPSHOT.jar.zip)

Also-by: Samuel Leisering <samuel@sleisering.de>
Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
@htreu @kaikreuzer With my latest commit I added the documentation. This finalizes my work and the code is ready for a review.

How should I handle the ip-validation issues?

**Comment by htreu** [comment]
> How should I handle the ip-validation issues?

I did the check manually, ECA is valid.

File: extensions/binding/org.eclipse.smarthome.binding.hue/ESH-INF/thing/LightLevelSensor.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+			<channel id="illuminance" typeId="illuminance" />
+			<channel id="light_level" typeId="light_level" />
+			<channel id="dark" typeId="dark" />
+			<channel id="daylight" typeId="daylight" />
+			<channel id="last_updated" typeId="last_updated" />
+			<channel id="battery_level" typeId="system.battery-level" />
+			<channel id="battery_low" typeId="system.low-battery" />
+		</channels>
+
+		<representation-property>uniqueId</representation-property>
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/ESH-INF/thing/LightLevelSensor.xml]: Shouldn't this be `sensorId`? Same for the other types.

File: extensions/binding/org.eclipse.smarthome.binding.hue/ESH-INF/thing/LightLevelSensor.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+		<config-description>
+			<parameter name="sensorId" type="text">
+				<label>Sensor ID</label>
+				<description>The identifier that is used within the hue bridge.</description>
+				<required>true</required>
+			</parameter>
+			<parameter name="on" type="boolean">
+				<label>Sensor Status</label>
+				<description>Indicates whether the sensor is enabled or not.</description>
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/ESH-INF/thing/LightLevelSensor.xml]: Does "indicate" mean that it is only for information or can we disable the sensor by changing this configuration?

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueBindingConstants.java [hunk 3]
```diff
@@ -22,7 +22,7 @@
  * @author Jochen Hiller - Added OSRAM Classic A60 RGBW
  * @author Markus Mazurczak - Added OSRAM PAR16 50
  * @author Andre Fuechsel - changed to generic thing types
- *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueBindingConstants.java]: Please add two lines, one for each author (ok to have the same description for both)

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java [hunk 4]
```diff
@@ -274,6 +374,24 @@ private synchronized void onUpdate() {
                 }
                 pollingJob = scheduler.scheduleWithFixedDelay(pollingRunnable, 1, pollingInterval, TimeUnit.SECONDS);
             }
+            if (sensorPollingJob == null || sensorPollingJob.isCancelled()) {
+                int pollingInterval = DEFAULT_SENSOR_POLLING_INTERVAL;
+                try {
+                    Object pollingIntervalConfig = getConfig().get(SENSOR_POLLING_INTERVAL);
+                    if (pollingIntervalConfig != null) {
+                        pollingInterval = ((BigDecimal) pollingIntervalConfig).intValue();
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java]: Are we sure that we can cast here? Does the framework make sure that `getConfig()` always converts the value to a BigDecimal or could it potentially be a String here?

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java [hunk 5]
```diff
@@ -274,6 +374,24 @@ private synchronized void onUpdate() {
...
+                    } else {
+                        logger.info(
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java]: debug should be enough

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java [hunk 6]
```diff
@@ -274,6 +374,24 @@ private synchronized void onUpdate() {
...
+                                "Sensor polling interval not configured for this hue bridge. Using default value: {}ms",
+                                pollingInterval);
+                    }
+                } catch (NumberFormatException ex) {
+                    logger.info("Wrong configuration value for sensor polling interval. Using default value: {}ms",
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java]: reduce to debug.

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueClient.java [hunk 7]
```diff
@@ -10,18 +10,20 @@
...
 import org.eclipse.smarthome.binding.hue.internal.FullLight;
+import org.eclipse.smarthome.binding.hue.internal.FullSensor;
 import org.eclipse.smarthome.binding.hue.internal.StateUpdate;
 
 /**
  * Access to the Hue system for light handlers.
  *
  * @author Simon Kaufmann - initial contribution and API
- *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueClient.java]: two lines

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueSensorHandler.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract Sensor Handler
+ *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueSensorHandler.java]: two lines, first should be "Initial contribution"

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueSensorHandler.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        initializeThing((bridge == null) ? null : bridge.getStatus());
+    }
+
+    private void initializeThing(@Nullable ThingStatus bridgeStatus) {
+        logger.debug("initializeThing thing {} bridge status {}", getThing().getUID(), bridgeStatus);
+        final String configSensorId = (String) getConfig().get(SENSOR_ID);
+        if (configSensorId != null) {
+            sensorId = configSensorId;
+            // note: this call implicitly registers our handler as a listener on
+            // the bridge
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueSensorHandler.java]: remove the line break

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/DimmerSwitchHandler.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.smarthome.binding.hue.internal.handler.HueSensorHandler;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+/**
+ * Hue Dimmer Switch
+ *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/DimmerSwitchHandler.java]: two lines

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/LightLevelHandler.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.unit.SmartHomeUnits;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+/**
+ * Light Level Sensor
+ *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/LightLevelHandler.java]: two lines

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/PresenceHandler.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import org.eclipse.smarthome.binding.hue.internal.handler.HueSensorHandler;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+/**
+ * Presence Sensor
+ *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/PresenceHandler.java]: two lines

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/TemperatureHandler.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+/**
+ * Temperature Sensor
+ *
+ * @author Samuel Leisering & Christoph Weitkamp - Added support for sensor API
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/sensors/TemperatureHandler.java]: two lines

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java (hunk 4 above)
Comment (cweitkamp) [reply extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/handler/HueBridgeHandler.java]: The nicest way to make sure we do not run into an exception would be to introduce a dedicated configuration class for it and use the `getConfigAs()` method. I will change it, if possible.

File: extensions/binding/org.eclipse.smarthome.binding.hue/ESH-INF/thing/LightLevelSensor.xml (hunk 1 above)
Comment (kaikreuzer) [reply extensions/binding/org.eclipse.smarthome.binding.hue/ESH-INF/thing/LightLevelSensor.xml]: Hm, wait. I just noticed that there is indeed a property on the Things called "uniqueId", which looks a bit like a MAC. So I guess if this also exists for sensors, we can keep it as is. But please check whether this is REALLY unique as we have decided to split a single sensor device into several functional Things - so there's some risk that they all have the same (device) unique id...

**Comment by cweitkamp** [comment]
@kaikreuzer Rebased and squashed.

**Comment by kaikreuzer** [comment]
Cool, thanks. I have created CQ 18371 - please refrain from any further changes to this PR!

**Comment by kaikreuzer** [comment]
Hooray, we have check-in approval!

**Comment by michaeljoos72** [comment]
@cweitkamp This integration is great, Thx! One proposal: It seems that temperature sensors - and I've read this as well in other communities - are not calibrated. My 5 sensors always show a temperature that is 1-2 degrees lower compared to my Netatmo. Would it be possible to integrate an offset to the temperature channel? BR Michael

**Comment by cweitkamp** [comment]
@michaeljoos72 I can confirm  that. You might  want to have a look into the [OffsetProfile](https://www.eclipse.org/smarthome/documentation/development/bindings/profiles.html#offsetprofile) to handle it.

**Pull request merged**